### PR TITLE
Image minification

### PR DESCRIPTION
**Pull request by duggiemitchell**
## Cute animal pic
_Because everyone likes pictures of animals._
little-big-dog

## Links
[Trello card](https://trello.com/c/EamUpG9W)

## Description
- All (I think) images should be resized and have `srcset` attributes

## Steps to test/reproduce
_Please explain how to best reproduce the issue and/or test the changes locally._

## Show me
Screenshot from 2020-03-31 18-31-54

---

## Reviewer tasks:

- [ ] Pulled branch and ran code locally (or viewed on preview deployment, if applicable)
- [ ] Reproduced the issue to review in the browser
- [ ] Reviewed code
- [ ] Updated Trello card (if applicable)

### CODE REVIEW

**Comment by mirisuzanne** [comment]
[I want automated responsive images & thumbnails](https://trello.com/c/EamUpG9W/44-i-want-automated-responsive-images-thumbnails)

File: content/_includes/embed.macros.njk [hunk 1]
```diff
@@ -323,7 +323,13 @@ params:
 ) -%}
 {%- if src -%}
 {%- set src = src if ('://' in src) else ('/assets/images/' + src) -%}
-<img src="{{ src }}" alt="{{ alt }}" {{ utility.show_attrs(attrs) }}/>
+<img
+  srcset="{{src | imgSuffix('320')}} 320w, {{src | imgSuffix('550')}} 550w"
+  sizes= "(max-width: 550px) calc(100vw - 120px), 550px"
+  src="{{ src }}"
```
Comment (mirisuzanne) [inline content/_includes/embed.macros.njk:329]: In the MDN examples, they duplicate the default image in both places. I'm going to look into that a bit, and see if it's necessary… 

I also wonder if we want to define our largest size as well (assuming many images will be larger than `550` by default), maybe something like `1200`?

Comment (mirisuzanne) [reply content/_includes/embed.macros.njk:329]: ok, Eric Portis pointed me to [the spec](https://html.spec.whatwg.org/multipage/images.html#introduction-3) which says that `src` is ignored when using `w`. So I think we do need to represent larger images in the `srcset` somehow.

Comment (duggiemitchell) [reply content/_includes/embed.macros.njk:329]: I think using `src` is helpful for unsupported browsers. Creating default image size at 1200 make sense.  I'm (doubtful, but) curious if this affects images that are naturally smaller. Would it load images that are too big for what we need?

File: src/filters/type.js [hunk 2]
```diff
@@ -8,14 +8,40 @@ const removeMd = require('remove-markdown');
...
+        {
+          width: 550,
+          rename: {
+            suffix: '-550',
+          },
+        },
+      ],
+    },
+    sizes: {
+      '*': '(max-width: 550px) calc(100vw - 120px), 550px',
```
[the remark this answers is not in the document]
Comment (duggiemitchell) [reply src/filters/type.js:32]: Sure, I can play with that, unless we decide leaving out `sizes` altogether and letting the browser make that choice instead.

File: content/_includes/embed.macros.njk (hunk 1 above)
Comment (mirisuzanne) [reply content/_includes/embed.macros.njk:329]: - Yeah, we don't want to remove `src`, but we do need to duplicate that path in the `srcset`
- If possible, I think we should only shrink images, and never enlarge them.

File: gulpfile.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+const gulp = require('gulp');
+const responsive = require('gulp-responsive');
```
Comment (jgerigmeyer) [inline gulpfile.js:3]: I'm open to this if it's the only way we can make it work... but I'm hesitant to add `gulp` unless we have to. I'm wondering if we could use something like https://github.com/MarcusCemes/responsive-image-builder to do the same thing? I'd have to dig into it a bit more.

**Comment by duggiemitchell** [comment]
@mirisuzanne I think I made the right fixes for your re-review. 

Note that I found something weird with the markdown plugin that we are using; it errs if there is no key for sizes (my assumptions were omissions are OK), so that key is left empty in our config.

File: gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+gulp.task('images', () =>
+  gulp
+    .src(`src/images/**/*.{jpg,png}`)
+    .pipe(
+      responsive(
+        {
+          // Resize all JPG images to three different sizes: 320, 550, and 1200 pixels
+          '**/*': [
```
[the remark this answers is not in the document]
Comment (duggiemitchell) [reply gulpfile.js:12]: oops, @mirisuzanne looks like you took care of this here: 742aa741b7d8702587e4bd16080607cbec73841a

Is this good to go, or are there other changes needed around this (other than fixing tests)?

[the remark this answers is not in the document]
Comment (mirisuzanne) [reply gulpfile.js:12]: @duggiemitchell I took care of it in the js filters, but not in the gulpfile. It was being hard-coded in both places. I wasn't sure if the same solution would work for the gulpfile. Can you take a look at that one still?

File: gulpfile.js (hunk 3 above)
Comment (duggiemitchell) [reply gulpfile.js:3]: I vote for merging this and seeking alternatives in another card for later; but could defer to @jgerigmeyer 's thoughts on that.

File: content/_includes/embed.macros.njk [hunk 5]
```diff
@@ -320,18 +341,34 @@ params:
...
   src,
   alt='',
+  width=none,
+  height=none,
   attrs={}
 ) -%}
 {%- if src -%}
 {%- set src = src if ('://' in src) else ('/assets/images/' + src) -%}
-<img src="{{ src }}" alt="{{ alt }}" {{ utility.show_attrs(attrs) }}/>
+{% set set = srcset(src, width) %}
```
Comment (jgerigmeyer) [inline content/_includes/embed.macros.njk:363]: whoa -- setting a variable to the name "set" seems like asking for trouble 🙈

File: content/_includes/embed.macros.njk [hunk 6]
```diff
@@ -309,6 +315,21 @@ params:
 {%- endif -%}
 {%- endmacro %}
 
+{% macro srcset(
+  src,
+  width=none
+) -%}
+  {%- set width = width or 640 -%}
```
Comment (jgerigmeyer) [inline content/_includes/embed.macros.njk:322]: @mirisuzanne Any reason to do this vs just setting `width=640` above?

File: content/_includes/embed.macros.njk [hunk 7]
```diff
@@ -309,6 +315,21 @@ params:
...
+  {%- set sizes = 'srcset' | fromTaxonomy -%}
+  {%- if width > sizes[0] -%}
+    {%- for size in sizes -%}
+      {%- if size < width -%}
+        {{- src | imgSuffix(size | string) }} {{ size + 'w,' -}}
+      {%- endif -%}
+    {%- endfor -%}
+    {{- src }} {{ width + 'w' -}}
+  {%- endif -%}
+{%- endmacro %}
```
Comment (jgerigmeyer) [inline content/_includes/embed.macros.njk:332]: @mirisuzanne Is it intentional that we're only using this in `figure()` and `post.hero()` images?

File: content/_includes/embed.macros.njk (hunk 6 above)
Comment (mirisuzanne) [reply content/_includes/embed.macros.njk]: I'm not actually sure… Do defaults work if the calling location passes `none` as the value?

Comment (jgerigmeyer) [reply content/_includes/embed.macros.njk]: Yes, the default value is used in place of `none`.

File: content/_includes/embed.macros.njk (hunk 7 above)
Comment (mirisuzanne) [reply content/_includes/embed.macros.njk:331]: I don't know what you mean. `post.hero()` does call `embed.img()`, which applies `srcset()`. I'm seeing srcsets on all relevant images when I look at the deploy preview. What am I missing?

Comment (mirisuzanne) [reply content/_includes/embed.macros.njk:331]: We use `embed.img` all over the site. Do you just mean that not all images have the height/width set? I figured this was worth merging before we take the time to update every single image.

Comment (mirisuzanne) [reply content/_includes/embed.macros.njk:331]: I've been adding dimension to every image I come across as I work, but we'll need a card to do it more intentionally

File: src/filters/type.js [hunk 8]
```diff
@@ -12,12 +12,14 @@ const markdownItResponsive = require('@gerhobbelt/markdown-it-responsive');
...
-  },
-}));
+const responsiveConfig = taxonomy.srcset
+  .map((size) => ({
+    width: size,
+    rename: {
+      suffix: `-${size}`,
+    },
+  }))
+  .concat([{}]);
```
Comment (jgerigmeyer) [inline src/filters/type.js:22]: @mirisuzanne @duggiemitchell This empty object means that the fallback original image is also processed and passed through along with the new resized ones.

Comment (mirisuzanne) [reply src/filters/type.js:22]: That's great, but without supplying the width, it results in[ invalid `srcset`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-srcset):

- "If no descriptor is specified, the source is assigned the default descriptor of 1x."
- "It is incorrect to mix width descriptors and pixel density descriptors in the same srcset attribute."

I'd love to talk through some of the isseues here, around supplying a width to markdown. Or we can try to move away from md images entirely to using `embed.img()` macro.

Comment (jgerigmeyer) [reply src/filters/type.js:22]: I don't follow... All this does is process the default images through the optimizer as well -- this change shouldn't impact the `srcset` attr at all.

Comment (mirisuzanne) [reply src/filters/type.js:22]: well maybe it's a problem with the plugin then? The result of a md image is not currently valid:

```html
<img src="/assets/images/blog/2016/cssconf/sd.jpg" alt="Sarah Drasner speaking" srcset="/assets/images/blog/2016/cssconf/sd-480.jpg 480w, /assets/images/blog/2016/cssconf/sd-960.jpg 960w, /assets/images/blog/2016/cssconf/sd-1280.jpg 1280w, /assets/images/blog/2016/cssconf/sd-2240.jpg 2240w, /assets/images/blog/2016/cssconf/sd.jpg ">
```

File: content/_includes/embed.macros.njk (hunk 6 above)
Comment (mirisuzanne) [reply content/_includes/embed.macros.njk:322]: It looks like you changed this, and changed it back?

Comment (jgerigmeyer) [reply content/_includes/embed.macros.njk:322]: Yeah, it works with `none`, but apparently not with `""`. 🤷

File: content/_data/taxonomy.yaml [hunk 9]
```diff
@@ -27,3 +27,17 @@ oss:
...
     show: contributor
+img:
+  # The size to use in src attribute, when srcset is not supported
+  fallback: 1280
+  # Sizes to generate automatically
+  srcset:
+    - 480
+    - 960
+    - 1280
+    - 2240
```
[the remark this answers is not in the document]
Comment (mirisuzanne) [reply content/_data/taxonomy.yaml:38]: I figured our industry has a lot of large "4k" or at least `2x` density screens, where `2240`  would be used for `1120px` width?

**Pull request merged**